jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -421,14 +421,16 @@ object XGBoost extends Serializable {
   }
 
   private def aggByGroupInfo(trainingData: RDD[XGBLabeledPoint]) = {
-    val normalGroups: RDD[Array[XGBLabeledPoint]] = trainingData.mapPartitions(
+    val baseGroups: RDD[XGBLabeledPointGroup] = trainingData.mapPartitions(
```

Comment:
According to its function, no mater how many groups in one partition, it is a big cost. I do not have large data with group, so I only use the data from "XGBoostRegressorSuite.ranking:use group data", and do 100000 times loops only for creation of normalGroups and edgeGroups between after and before add this commit, here is the performance situation:
before add commit:  383507(millis),  388359(millis)
after add commit: 324289(millis), 322539(millis)

Reply:
I am actually very curious how you can `only creation of normalGroups and edgeGroups `

as I said RDD is lazily evaluated, if you don't run an actual job Spark does nothing

if you run a job, it only takes 6 minutes to run 100000 Spark jobs?

Reply:
But, it is a small data from "XGBoostRegressorSuite.ranking:use group data", In the real situation, it will be a large scale data. I do lots of this experiments, if you add this lazy formation, the cost time will continues enhance, So, I do not thinks it is some magic of JIT

Reply:
looks like you have a wrong understanding on basic mechanism of Spark, see my previous comments

Reply:
I think you are `benchmarking` JVM performance by creating 100000 additional variables........

Reply:
you can add performance check logic in aggByGroupInfo, and run "XGBoostRegressorSuite.ranking:use group data"

Reply:
if you want to discuss, please read my comments and check if you missed something in understanding mechanism

not push back!

Reply:
I think we have a understand gap, sure, you are right, spark mechanize is lazy model, and the bench marking is JVM performance by creating 100000 additional varibales, but as I say, we only use small data, and I do think it is a good code format

Reply:
....you are creating a `actually not exist` loop to create a `only exist in your code` performance overhead to say you improved the performance 

it does not make any sense to me

Reply:
this additional variables are created by you and it's not related to data size,

Reply:
I use a large data about two million, and do a end-to-end test, there is not a big enhancement. Later I will use more real data to do my work before adding commit, anyway thanks for your time